ERCS/erc-7811.md

```diff
@@ -0,0 +1,222 @@
+---
+
+## Abstract
+
+This ERC introduces a new RPC call, wallet_getAssets, for wallets to declare to the Dapp what assets are owned by the user. This allows for more accurate asset discovery and the use of assets that aren’t available on-chain but can be provided by the wallet
+
+## Motivation
+
+Currently, Dapps primarily rely on on-chain data to determine a user's balance, which can be limiting. Furthermore, a Dapp might restrict the user from initiating actions that the wallet could otherwise resolve, as it cannot account for the total assets a user has across different accounts or chains.
+
+Wallets already have information about a user's assets, including those not visible on-chain, and need a way to communicate that information to Dapps.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Method: `wallet_getAssets`
+
+#### Request schema
+
+```ts
+type WalletGetAssetsRequest = {
+  account: Hex;
+  requiredAssets?: Record<Hex, Hex[]>;
+};
+```
+
+`account` is a required field that indicates for which account assets are requested.
+
+`requiredAssets` is an optional field that specifies only the assets Dapp cares about on the specific chains. If it is provided, the response from the wallet SHOULD include those assets.
+
+#### Example request
+
+```json
+{
+  "account": "0x123",
+  "requiredAssets": {
+    "0x1": ["0x456", "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"],
+    "0xa": ["0x789", "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"]
+  }
+}
+```
+
+#### Response schema
+
+```ts
+type Asset = {
+  address: Hex;
+  balance: Hex;
+  type: string;
+  metadata: any;
+};
+type WalletGetAssetsResponse = Record<Hex, Asset[]>;
+```
+
+The key **SHOULD** be [EIP-155](./eip-155.md) chainId
+
+Asset fields:
+
+`address` is the address of the asset as Hex. Native assets **MUST** use `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` address specified by [ERC-7528](./eip-7528.md)
+
+`balance` is the balance of the asset as Hex
+
+**`type`:** A string indicating the type of the asset. Common asset types include but **aren’t limited to**:
+
+- **`ERC20`:** For [ERC-20](./eip-20.md) tokens
+- **`ERC721`:** For [ERC-721](./eip-721.md) tokens (NFTs)
+- **`native`:** For the chain's native currency
+
+**`metadata`:** An optional object containing additional information about the asset. The specific fields within the metadata object may vary depending on the asset type and the wallet's implementation."
+
+#### Example response
+
+```json
+{
+  "0x1": [
+    {
+      address: "0x123",
+      balance: "0xcaaea35047fe5702",
+      type: "ERC20",
+      metadata: {
+        name: "Token",
+        symbol: "TOK",
+        decimals: 18,
+      }
+    },
+    {
+      address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
+      balance: "0xcaaea35047fe5702",
+      type: "native"
+    }
```

Comment:
Wonder if we should have a rule that native asset must be the first item in the array. Would feel a bit weird to have a linear time complexity just to extract a native balance (and there could be many many tokens).

Reply:
Would [this suggestion](https://github.com/ethereum/ERCs/pull/709/files/ff4b3eb3ac8e275d5f055546caed66919ab0d3d0#r1853008255) resolve your concern?